A59-audit-logging.md

```diff
@@ -0,0 +1,243 @@
+A59: gRPC Audit Logging
+----
+* Author(s): [Luwei Ge](https://github.com/rockspore)
+* Approver:
+* Status: Draft {Draft, In Review, Ready for Implementation, Implemented}
+* Implemented in:
+* Last updated: 2023-03-09
+* Discussion at: 
+
+## Abstract
+
+Support audit logging based on authorization events in gRPC servers in both
+xDS-enabled and gRPC authorization policy based use cases.
+
+## Background
+
+gRPC's authorization support is based on the [RBAC policy][]. In
+xDS-enabled cases, users can configure authorization policies in the control
+plane which applies [RBAC HTTP filters][RBAC filter] to the gRPC servers. When
+xDS is not an option, users can use the [gRPC authorization policy][A43] to
+get similar authorization enforment.
+
+This proposal describes the audit logging APIs that allow users to audit
+specific authorization events in their gRPC servers with the necessary context.
+We aim to offer consistent APIs to support both aforementioned use cases.
+
+[RBAC filter]: https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/http/rbac/v3/rbac.proto
+[RBAC policy]: https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/rbac/v3/rbac.proto
+[HttpConnectionManager proto]: https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
+
+### Related Proposals: 
+* [gRFC A41: xDS RBAC Support][A41]
+* [gRFC A43: gRPC Authorization API][A43]
+
+[A41]: A41-xds-rbac.md
+[A43]: A43-grpc-authorization-api.md
+
+## Proposal
+
+### Audit Context
+
+The table below lists the metadata we will include in the audit context,
+which will be made available to audit loggers during an audit event.
+
+It is worth noting that audit happens right after the authorization
+decision is made, so the status only reflects that but not the eventual
+one from the server. 
+
+|Fields             |Additional Information                                                             |Example                |
+|-------------------|-----------------------------------------------------------------------------------|-----------------------|
+|RPC Method         |Method the RPC hit.                                                                |"/pkg.service/foo"     |
+|Principal          |Identity the RPC. Currently only available in certificate-based TLS authentication.|"spiffe://foo/user1"   |
+|Timestamp          |Time when the RPC happens.                                                         |1649376211             |
+|Denying Policy Name|Policy name that denied the RPC. Empty if RPC is allowed.                          |"policy_name"          |
```

Comment:
Isn't it possible that we also have a particular policy name response for allowing a request, if we're using an RBAC filter with an ALLOW action?

Also, aren't there possible configurations where we'll wind up generating multiple log entries for each RPC, one for each RBAC filter?  Don't we need some way to know that those duplicate messages all refer to the same RPC?  In particular, note that there is no restriction that prevents two RBAC filters from using the same policy name.

Reply:
First of all, given the scenario you used as as example, I realized it's still necessary to add some more explanation about the `ON_DENY` and `ON_ALLOW`. Basically, I think it'd be quite intuitive to consider `ON_DENY` condition met when a RBAC rejects the RPC, even if the action of it is `ALLOW`. Likewise, I think we should also consider `ON_ALLOW` condition met when a RBAC lets the RPC through, even if the action could be `DENY`. If the overloading of `DENY` and `ALLOW` for action and decision here could cause confusion, we can think about using other words like `ON_AUTHORIZED` or so. WDYT?

About logging the policy name, I changed it to log the allowing policy name as well for now. But I think some discussion related to this is still needed:
1. Like you said, one caveat is when there could be multiple log entries. I believe we discussed it before where we assumed this was not what users would generally want. In the gRPC authz policy section, I specifically made this assumption and added more paragraphs to explain the configuration. PTAL.
2. In xDS, particularly where users can have arbitrary numbers of RBAC filters, they could audit the same RPC multiple times if they really want. Let me know if I am wrong but I did't find gRPC generating any sort of request UUID on the server side to identify a RPC. So if users really want to do this, they would need to think about how to dedup the audits if needed. I guess if feasible, we could include some RPC metadata which users could configure out a way with.
3. Now let's assume users only want to audit once even with multiple RBACs, so they would only configure `ON_ALLOW` at the last RBAC. The question is how meaningful it would be to log just the policy name of the last RBAC? I don't think we should simply accumulate policy names here for audit on allow.

I don't know where users would reasonably want to use the same policy name in multiple active RBAC filters, _and_ also want to audit log multiple times. I'd prefer we consider this as an anti-pattern and I don't see much we could do to resolve the confusion.

Reply:
1. I don't think it will be a common config, but it is a possible config, so we need to consider the implications when that config is used.
2. Correct, there is no UUID for each RPC.  And I agree that that's an issue for the multiple-logging case.  I suppose we could try to generate some sort of UUID, although I'm not crazy about it.
3. I don't fully follow the case you're describing here.  In general, I think that if there is a particular policy name, it seems right to log it.  But I think there may be cases where they would not be a specific policy name to log: for example, if an RBAC policy has action `DENY` but has audit condition `ON_ALLOW`, then we will log a message if none of the policies match.

I'd like to get @ejona86's thoughts on this.

Reply:
In #3, what I meant was if there are two RBACs present, an authorized request would need to be allowed by both of them. Now if users only want one log entry on allow, which is the assumption we have in SDK, the audit condition `ON_ALLOW` will only be added to the second RBAC policy. It will not be accurate to only log the second RBAC policy name in this case though, since the first RBAC technically contributed to allowing the request as well.

Reply:
During our offline discussion, we agreed that it was a reasonable expectation to audit at most once per RPC and the gRPC authorization API (SDK) makes this assumption for the users. We also concluded that while the possibility of configuring multiple audit logs per RPC is acknowledged, we would not consider providing any mechanism to de-duplicate those log entries in this gRFC.

As suggested, I also added more paragraphs elaborating on this in the `xDS API changes` section + `Audit Condition in the HTTP Connection Manager` section in `Rationale`. PTAL. @markdroth @ejona86

Reply:
I think that the policy name is useful in the DENY case, but questionable in the ALLOW case.  If you were outputting policy names for ALLOW, you would want the policy names of all ALLOW rules that had matched, which doesn't seem worth the complication of supporting.  If there is only 1 ALLOW rule, then the name doesn't need to be logged since it must always be that rule's policy.

Reply:
To avoid confusion, the DENY and ALLOW cases in your comment above refer to the RBAC actions, not decisions. 

According to the definition in [RBAC](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/rbac/v3/rbac.proto), "A match occurs when at least one policy matches the request. The policies are evaluated in lexicographic order of the policy name.", so the behavior is actually deterministic regarding which policy (aka rule in SDK) gets matched. So I think it's reasonable to just log that policy name.